modules/transactions/src/testFixtures/java/org/apache/ignite/internal/tx/storage/state/AbstractTxStateTableStorageTest.java

```diff
@@ -0,0 +1,170 @@
+
+        tableStorage.startRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS);
+
+        TxStateStorage newStorage0 = tableStorage.getTxStateStorage(PARTITION_ID);
+
+        assertNotNull(newStorage0);
+
+        assertSame(storage, newStorage0);
+
+        assertEquals(0L, newStorage0.lastAppliedIndex());
+        assertEquals(0L, newStorage0.persistedIndex());
+
+        assertThat(getAll(newStorage0.scan()), empty());
+
+        tableStorage.startRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS);
+
+        TxStateStorage newStorage1 = tableStorage.getTxStateStorage(PARTITION_ID);
+
+        assertSame(newStorage0, newStorage1);
+    }
+
+    @Test
+    public void testAbortRebalance() throws Exception {
+        assertDoesNotThrow(() -> tableStorage.abortRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS));
+
+        UUID oldTxId = UUID.randomUUID();
+        TxMeta oldTxMeta = mock(TxMeta.class);
+
+        TxStateStorage storage = tableStorage.getOrCreateTxStateStorage(PARTITION_ID);
+
+        storage.put(oldTxId, oldTxMeta);
+
+        storage.lastAppliedIndex(100);
+
+        tableStorage.startRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS);
+
+        UUID newTxId = UUID.randomUUID();
+        TxMeta newTxMeta = mock(TxMeta.class);
+
+        storage.put(newTxId, newTxMeta);
+
+        storage.lastAppliedIndex(500);
+
+        tableStorage.abortRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS);
+
+        assertSame(storage, tableStorage.getTxStateStorage(PARTITION_ID));
+
+        assertDoesNotThrow(() -> tableStorage.abortRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS));
+
+        assertEquals(100L, storage.lastAppliedIndex());
+
+        assertThat(getAll(storage.scan()), containsInAnyOrder(new IgniteBiTuple<>(oldTxId, oldTxMeta)));
+    }
+
+    @Test
+    public void testFinishRebalance() throws Exception {
+        assertDoesNotThrow(() -> tableStorage.finishRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS));
+
+        UUID oldTxId = UUID.randomUUID();
+        TxMeta oldTxMeta = mock(TxMeta.class);
+
+        TxStateStorage storage = tableStorage.getOrCreateTxStateStorage(PARTITION_ID);
+
+        storage.put(oldTxId, oldTxMeta);
+
+        storage.lastAppliedIndex(100);
+
+        tableStorage.startRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS);
+
+        UUID newTxId = UUID.randomUUID();
+        TxMeta newTxMeta = mock(TxMeta.class);
+
+        storage.put(newTxId, newTxMeta);
+
+        storage.lastAppliedIndex(500);
+
+        tableStorage.finishRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS);
+
+        assertSame(storage, tableStorage.getTxStateStorage(PARTITION_ID));
+
+        assertDoesNotThrow(() -> tableStorage.finishRebalance(PARTITION_ID).get(1, TimeUnit.SECONDS));
+
+        assertEquals(500L, storage.lastAppliedIndex());
+
+        assertThat(getAll(storage.scan()), containsInAnyOrder(new IgniteBiTuple<>(newTxId, newTxMeta)));
+    }
+
+    private static <T> List<T> getAll(Cursor<T> cursor) {
```

Comment:
Why not just throw an exception? See no point in wrapping it into a runtime exception

Reply:
Fix it